src/firewall/core/rich.py

```diff
@@ -433,104 +435,69 @@ def check(self, family=None):
 }
 
 
+@dataclass(frozen=True)
 class Rich_Limit(_Rich_Entry):
-    def __init__(self, value, burst=None):
-        self.value = value
-        self.burst = burst
+    """ This object only holds data and is read-only after init. It is also
+    hashable and can be used as a dictionary key. """
+    value: str = field(compare=False) # init only, use rate and duration
+    burst: int = None
+    rate: int = field(init=False)
+    duration: str = field(init=False)
 
-    def check(self, family=None):
-        self.value_parse()
-        self.burst_parse()
+    def __post_init__(self):
+        rate, duration = self._value_parse()
+        object.__setattr__(self, "rate", rate)
+        object.__setattr__(self, "duration", duration)
```

Comment:
now that calling `Rich_Limit()` can fail with an exception, did you check all callers that they handle that correctly?

Reply:
`self.value` does not, but we also don't use it. It's only used to set `self.rate` and `self.duration`.

`self.burst` is verified to be an int and range. What else needs to be done to "normalize" it ?

Reply:
> now that calling Rich_Limit() can fail with an exception, did you check all callers that they handle that correctly?

No. I didn't intend my commit to be complete; just illustrating other options.
I'm guessing that in most cases we want to propagate the error, i.e. it came from the UI or startup.

Reply:
Fixed remaining usage of `limit.value`.

Reply:
> self.value does not, but we also don't use it

Isn't it used for the hashing/equality implementation?

There are probably various places, that do care about the string representation of the value (preferably, in a normalized, expected form). For example, `ip4tables._rule_limit()`.

Reply:
I'd actually like to consider making `limit.value` throw an exception when it's read. That would make sure nothing uses it. :)

Reply:
I think it's sensible that `ip4tables._rule_limit()` uses `limit.value`, and doesn't have to go through to construct it from `f"{limit.rate}/{limit.duration}"`.

Sure, that means, `iptables` backend is fine with the default string format that `Rich_Rule.value` provides. Other users may not (nftables), and may need to implement their own thing based on the more fundamental values.

But leaving `value` unnormalized there, and then require `ip4tables._rule_limit()` to reimplement it, is not useful. It also means, that the constructed rule contains a newly allocated string, instead of sharing the one from `Rich_Rule.value`.

Reply:
in other words, `value` has a sensible form to which we can normalize. And "coincidentally", that is also what `ip4tables._rule_limit()` needs. Win-Win.

Reply:
pushed one line change to normalize `self.value`. I would still prefer ipXtables to not rely on this.